ASSIGNMENT 1

Group 67

300145970 Lixiong Wei

300126813 Zhiyuan Lin

**Knapsack Problem**

**Import Dataset**

The dataset used is for knapsack problem, which determine the optimal way to store item into bag
with minimal weight and maximal value.

Dataset Reference: https://www.kaggle.com/datasets/warcoder/knapsack-problem?resource=download


In [1]:
#Import libraries needed

import pandas as pd
import itertools
import numpy as np

**Read DataSet**

Raw data link: https://raw.githubusercontent.com/William-67/CSI4106/main/A1/knapsack_5_items.csv


In [11]:
url = 'https://raw.githubusercontent.com/William-67/CSI4106/main/A1/knapsack_5_items.csv'
dataset = pd.read_csv(url)


In [12]:
dataset.columns

Index(['Weights', 'Prices', 'Capacity', 'Best picks', 'Best price'], dtype='object')

**Test the data by printing out first 10 rows **

In [13]:
dataset.head(10)

Weights            Prices  Capacity        Best picks  Best price
0  [46 40 42 38 10]  [12 19 19 15  8]        40  [0. 1. 0. 0. 0.]        19.0
1  [11 31  4  6  7]  [ 2  8 18 16  3]        64  [1. 1. 1. 1. 1.]        47.0
2  [32 49 27 37 24]  [19 16 16  4  1]        87  [1. 0. 1. 0. 1.]        36.0
3  [20 35 22 23 16]  [19 17 19  9  1]        21  [1. 0. 0. 0. 0.]        19.0
4  [ 7 12 19 13 20]  [10 11 18 15  5]        50  [0. 1. 1. 1. 0.]        44.0
5  [27 10 25 25  7]  [13 19  7 16  3]        66  [1. 1. 0. 1. 0.]        48.0
6  [21  2 33 45 26]  [ 1 14 10  6 13]        80  [0. 1. 1. 0. 1.]        37.0
7  [37 27 39 14 25]  [18  7 15  4 13]        35  [0. 0. 0. 0. 1.]        13.0
8  [ 1 48  4 23 39]  [ 9  4 10 16 12]        51  [1. 0. 1. 1. 0.]        35.0
9  [ 4  3 22  9 32]  [14  6  3 17  8]        53  [1. 1. 0. 1. 1.]        45.0

Write a function to convert data into array of floats or integers

ex. [45 40 42 38 10] -> [45,40,42,38,10]

In [14]:
def string_to_list(string):

  string_list = string.strip('[]').split()

  float_list = [float(element) for element in string_list]

  return float_list

Furthermore, it's possible that certain rows in the dataset contain empty values in specific columns. We also aim to eliminate these rows as they do not provide any useful information. We use dropna() function to do so:


In [15]:
#Ignore the warning messages.

dataset = dataset.dropna()

dataset.Weights = dataset.Weights.apply(lambda x : string_to_list(x))
dataset.Prices = dataset.Prices.apply(lambda x : string_to_list(x))
dataset['Best picks'] = dataset['Best picks'].apply(lambda x : string_to_list(x))

<ipython-input-15-833fb421dc49>:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dataset.Weights = dataset.Weights.apply(lambda x : string_to_list(x))
<ipython-input-15-833fb421dc49>:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dataset.Prices = dataset.Prices.apply(lambda x : string_to_list(x))
<ipython-input-15-833fb421dc49>:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the docum

In [44]:
#This is a self test to view how the data in row looks like

for a, row in dataset.iterrows():
  if a<3:
    print(row)

Weights       [46.0, 40.0, 42.0, 38.0, 10.0]
Prices         [12.0, 19.0, 19.0, 15.0, 8.0]
Capacity                                  40
Best picks         [0.0, 1.0, 0.0, 0.0, 0.0]
Best price                              19.0
Name: 0, dtype: object
Weights       [11.0, 31.0, 4.0, 6.0, 7.0]
Prices        [2.0, 8.0, 18.0, 16.0, 3.0]
Capacity                               64
Best picks      [1.0, 1.0, 1.0, 1.0, 1.0]
Best price                           47.0
Name: 1, dtype: object
Weights       [32.0, 49.0, 27.0, 37.0, 24.0]
Prices          [19.0, 16.0, 16.0, 4.0, 1.0]
Capacity                                  87
Best picks         [1.0, 0.0, 1.0, 0.0, 1.0]
Best price                              36.0
Name: 2, dtype: object


**Generate and Test**

Generate all situation and choose the optimal one (**Maximum Price**) in each row

In [17]:
def gen_and_test(data):

  # initialize
  weights = data['Weights']
  prices = data['Prices']
  capacities = data['Capacity']

  best_solution_price = 0
  best_solution = [0,0,0,0,0]

  # the five for loop contains 2^5 = 32 situations

  for i in range(2):
    for j in range(2):
      for k in range(2):
        for m in range(2):
          for l in range(2):

            #calculating each weight and price
            weight = weights[0] * i + weights[1] * j + weights[2] * k + weights[3] * m + weights[4] * l
            price = prices[0] * i + prices[1] * j + prices[2] * k + prices[3] * m + prices[4] * l

            #hard constraint part
            if weight <= capacities:

              if price > best_solution_price:

                best_solution_price = price
                best_solution = [i,j,k,m,l]

  return best_solution_price, best_solution

In [18]:
solutions = []
for _, row in dataset.iterrows():
    target = row['Best price']
    solution, indexes = gen_and_test(row)
    solutions.append(1 if target == solution else 0)


In [19]:
# Accuracy
print('Accuracy of best prices found is', np.mean(solutions))

Accuracy of best prices found is 1.0


**Analysis:**

The generate and test algorithm gives us 1.0 accuracy, which is 100% match with the choice provided by the authors.

The algorithm is able to handle small amout of data. In this example, for each row of data, there are only 2^5 = 32 situations. Therefore; each situation can be evaluated which makes decision making easier.

--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

**Implement 3 more algorithm for this problem**

**1. Greedy Search**



Greddy search is an approach for solving a problem by selecting the best option available at the moment and doesn't worry if the final result is optimal.

There are two types of strategy to implement the greedy algorithm
1. go by **highest value**, for two equal values, choose the least heavy
2. go by **heaviest**. choose the highest value if two weight equal

We decided to use the first approach.

In [41]:
def greedy(data):

  #initialize
  weights = data['Weights']
  prices = data['Prices']
  capacity = data['Capacity']

  best_solution_price = 0
  best_solution = [0,0,0,0,0]

  #create a temperatory list for recording the prices and weights sorted in descending order
  #In order to maintain the integrity of origin list
  #In another words, Deep Copy

  tmp_prices = [0,0,0,0,0]
  tmp_weights = [0,0,0,0,0]
  for i in range(len(prices)):
    tmp_prices[i] = prices[i]
    tmp_weights[i] = weights[i]

  #start the choosing process
  curr_weight = 0

  #Keep filling the bag if room available
  #The process should stop when there's no item can be chosen
  while(curr_weight < capacity and len(tmp_prices)!=0):

    #choose the item with max value
    max_value = prices[0]
    item_index = 0

    for i in range(len(tmp_prices)):

      if max_value < tmp_prices[i]:
        max_value = tmp_prices[i]
        item_index = i

    if(curr_weight <= (capacity - tmp_weights[item_index])):

      best_solution_price += tmp_prices[item_index]
      best_solution[item_index] = 1
      curr_weight += tmp_weights[item_index]

    #remove the item after checking
    tmp_prices.remove(tmp_prices[item_index])
    tmp_weights.remove(tmp_weights[item_index])

  return best_solution_price, best_solution

In [42]:
solutions_greedy = []
for _, row in dataset.iterrows():
    target = row['Best price']
    solution, indexes = greedy(row)
    solutions_greedy.append(1 if target == solution else 0)


In [43]:
print("Greedy Accuracy is", np.mean(solutions_greedy))

Greedy Accuracy is 0.7880352262374735


**Analysis:**

The generate and test algorithm gives us 0.788 accuracy, which is 78.8% match with the choice provided by the authors.

The alogorithm is not as accuracy as generate and test because greedy search does not output the optimal solution. However, we obtain 78% accuracy, which is also consider accurate.

----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

**2. Simulated Annealing**

Simulated Annealing algorithm comes from physics, which is used to find the global optima. The algorithm set an initial state and keep track of temperature, then gentally lower down the temperature and decide the new optimal value based on if the exponential of the change is greater than random number from 0 to 1. The algorithm stop when meets with minimal requirement.

In [31]:
import random
import math


def simulated_annealing(data, N, initial_temperature, cooling_rate):

  #initialize
  weights = data['Weights']
  prices = data['Prices']
  capacity = data['Capacity']
  num_items = len(weights)

  # Function to calculate the total price of a solution
  def calculate_total_price(solution):
      total_price = 0
      total_weight = 0
      for i in range(num_items):
          total_price += solution[i] * prices[i]
          total_weight += solution[i] * weights[i]
      return total_price, total_weight

  # Initialize the current solution randomly
  current_solution = [0,0,0,0,0]
  current_price, current_weight = calculate_total_price(current_solution)

  # Initialize the best solution
  best_solution = current_solution.copy()
  best_solution_price, best_weight = calculate_total_price(best_solution)

  #list of founded neighbour solution so that would not go back again
  neighbour_list = []

  # Main simulated annealing loop
  temperature = initial_temperature
  #define the minimal temperature, this can be changed
  while temperature > 0.005:

      # Generate a neighboring solution by flipping a random item
      neighbor_solution = current_solution.copy()

      index_to_flip = random.randint(0, num_items - 1)
      neighbor_solution[index_to_flip] = 1 - neighbor_solution[index_to_flip]  # Flip 0 to 1 or 1 to 0
      neighbor_price, neighbor_weight = calculate_total_price(neighbor_solution)

      # Calculate the change in cost (price) due to the neighbor solution
      if not (neighbor_solution in neighbour_list):
        cost_change = neighbor_price - current_price

        # If the neighbor solution is better or accepted with a certain probability, update the current solution
        if cost_change > 0 or random.random() < math.exp(cost_change / temperature):
          current_solution = neighbor_solution
          current_price = neighbor_price
          current_weight = neighbor_weight

          # Update the best solution if needed
          if current_price > best_solution_price and current_weight <= capacity:
            best_solution = current_solution.copy()
            best_solution_price = current_price

          neighbour_list.append(current_solution)

        # Reduce the temperature
      temperature *= cooling_rate
  return best_solution_price, best_solution


In [36]:
solutions_sa = []
for _, row in dataset.iterrows():
    target = row['Best price']
    solution, indexes = simulated_annealing(row, N = 10, initial_temperature=1, cooling_rate=0.95)
    solutions_sa.append(1 if target == solution else 0)


**Changing the initial temperature has significant effect on the accuracy!**

**See the analysis for more detail**

In [37]:
print("Simulated Annealing Accuracy is", np.mean(solutions_sa))

Simulated Annealing Accuracy is 0.1947565543071161


**Analysis:**

Simulated Annealing algorithm is used to find global optima; however, it is not always correct. The accuracy depends on the comparsion result between the expoential of change and a random number. We try with larger number of iteration (ex. change the initial temperature), with larger num of iteration, the accuracy can be increased.


---

**Genetic Algorithm (Population-Based algorithm)**




**NOTICE**: we try to reduce the time of this algorithm but it still takes around **1 minutes** to run

In [38]:
import random


def calculate_fitness(ind, prices, weights, capacity):
    total_weight = sum(ind[i] * weights[i] for i in range(len(ind)))
    if total_weight > capacity:
      return 0  #solutions that exceed the capacity
    fitness = sum(ind[i] * prices[i] for i in range(len(ind)))
    return fitness

def crossover(parent1, parent2, cross_rate):

    if random.random() < cross_rate:

      #determine if the point where child
      crossover_point = random.randint(0, len(parent1))
      child1 = parent1[:crossover_point] + parent2[crossover_point:]
      child2 = parent2[:crossover_point] + parent1[crossover_point:]
    else:
        child1, child2 = parent1, parent2
    return child1, child2


def mutation(child, mut_rate):
    for i in range(len(child)):
        if random.random() < mut_rate:
            child[i] = 1 - child[i]  # Flip the bit (0 to 1 or 1 to 0)
    return child

def genetic_algorithm(data, population_size, num_generations, mut_rate, cross_rate, tournament_size):
    weights = data['Weights']
    prices = data['Prices']
    capacity = data['Capacity']
    num_items = len(weights)

    best_solution = [0,0,0,0,0]
    best_solution_price = 0


    # Generate the population randomly
    # might have duplicated solution
    population = [[random.randint(0, 1) for _ in range(num_items)] for _ in range(population_size)]

    for generation in range(num_generations):

        # Calculate fitness for each individual in the population
        fitness_values = [calculate_fitness(ind, prices, weights, capacity) for ind in population]

        # Select parents for the next generation using tournament selection
        # choose the best "tournament_size" parents
        selected_parents = random.sample(population, tournament_size)

        #Start the mutation process
        parent1, parent2 = random.sample(selected_parents, 2)
        child1, child2 = crossover(parent1, parent2, cross_rate)
        child1 = mutation(child1, mut_rate)
        child2 = mutation(child2, mut_rate)

        # Find the best solution in the current generation
        generation_best = child1
        generation_best_price = calculate_fitness(child1,prices,weights,capacity)
        if calculate_fitness(child1,prices,weights,capacity) < calculate_fitness(child2,prices,weights,capacity):

          generation_best_price = calculate_fitness(child2,prices,weights,capacity)
          generation_best = child2


        # Update the overall best solution if needed
        if generation_best_price > best_solution_price:
            best_solution = generation_best.copy()
            best_solution_price = generation_best_price

    return best_solution_price, best_solution

In [39]:
solutions_ga = []
for _, row in dataset.iterrows():
    target = row['Best price']
    solution, indexes = genetic_algorithm(row, population_size = 50, num_generations = 50, mut_rate = 0.1, cross_rate = 0.9, tournament_size = 5)
    solutions_ga.append(1 if target == solution else 0)


In [40]:
print("Genetic Algorithm Accuracy is", np.mean(solutions_ga))

Genetic Algorithm Accuracy is 0.929547525053143


**Analysis:**

Genetic algorithm takes longer than the other two algorithm, however; it's accuracy is the highest among 3 algorithms.

---

## **Comparison among Greedy Search, Simulated Annealing and Genetic algorithm:**

             Time(s)      Accuracy
Greedy_Search       1s          79%

Simulated_Annealing    4s          20%

Genetic Algorithm     71s         93%


Throught the given test, we find that the genetic algorithm we design spends the most time but the highest accuracy. The greedy search spend the least time aournd 1s to search the solution with 79% accuracy. The simulated Annealing algoritm seem is the worst method only has 20% accuracy which is the lowest one.



# impact of varying parameters on Simulated Annealing

Simulated Annealing:

initial temrature: 1 -> 20 -> 50
accuracy:      20% -> 55% -> 70%

Initial temprature determines how much the algorithm can explore the solution space initially. A high initial temperature allows the algorithm to accept more non-improving solutions initially, promoting exploration over exploitation. As the initial temprature increase, the solution amounts goes up slower.


Cool rate: 0.95 -> 0.5 -> 0.1
accuracy: 20% -> 11% -> 0.06%

The cooling rate significantly negatively affects the accuracy of the algorithm



# impact of varying parameters on Genetic algorithm:

Population Size:


Higher population Size generates more diverse solutions, leading to a better chance of finding the optimal or near-optimal solution.Increased computational cost due to larger population to handle.

Number of Generations:

More chances for the algorithm to improve the solutions and might contains over-iterations where no significant improvement is seen over multiple generations, wasting computational resources. Low generation amounts cause the inaccuracy.


Mutation Rate:

Higher mutation rate generates more randomness and explorations. If the rate is over-sized, it could disrupt good solutions, making the algorithm behave more like a random search. Low mutation rate cause stabilisation and refines the current solutions which the population might lack diversity, leading to premature convergence.

Crossover Rate:
High crossover rate generates more solutions in the population undergo crossover, producing more diverse offspring.The algorithm might rely too much on recombination and not enough on preserving good existing solutions.
Low crossover rate maintains the original population and the algorithm lose its function.



**Conclusion:**

With small amount of data, generate and test guarantee 100% accuracy

However; with larger amount, some algorithm should be deployed.

Greedy Search is to find the best solution compare to current situation, solution is not always the best but accuracy is acceptable

Simulated Annealing algorithm works for large number of iteration, the larger the temperature range, the more accurate the result is. However, if the range is not large enough, accuracy is low.

Genetic algorithm takes longer time compare to the the first two algorithms since the tournament lasts for many generation. Most of the result it produces are correct.
